libnmstate/nm/route.py

```diff
@@ -0,0 +1,112 @@
+#
+# Copyright 2019 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from operator import itemgetter
+from libnmstate.nm import nmclient
+from libnmstate.error import NmstateInternalError
+from libnmstate.schema import Route
+
+from libnmstate import iplib
+
+NM_MAIN_ROUTE_TABLE_ID = 0
+
+
+def get_routes(acs_and_ip_cfgs):
+    """
+    The acs_and_ip_cfgs should be generate to generate a tuple:
+        (NM.NM.ActiveConnection, NM.SettingIPConfig) for querying save config
+            or
+        (NM.NM.ActiveConnection, NM.IPConfig) for querying running routes
+    """
+    routes = []
+    for (active_connection, ip_cfg) in acs_and_ip_cfgs:
+        iface_name = _get_iface_name(active_connection)
+        if not iface_name:
+            raise NmstateInternalError(
+                'Got connection {} has not interface name'.format(
+                    active_connection.get_id()))
+        if _is_query_config(ip_cfg):
+            # Route table ID is stored in configure ipvx.route-table
+            table_id = _get_cfg_route_table_id(ip_cfg)
+        for nm_route in ip_cfg.props.routes:
+            if not _is_query_config(ip_cfg):
+                table_id = _get_running_route_table_id(nm_route)
+            route_entry = _nm_route_to_route(
+                nm_route,
+                table_id,
+                iface_name)
+            if route_entry:
+                routes.append(route_entry)
+    routes.sort(key=itemgetter(Route.TABLE_ID,
+                               Route.NEXT_HOP_INTERFACE,
+                               Route.DESTINATION))
+    return routes
+
+
+def _get_cfg_route_table_id(ip_profile):
+    table_id = iplib.KERNEL_MAIN_ROUTE_TABLE_ID
+    if ip_profile and ip_profile.props.route_table != NM_MAIN_ROUTE_TABLE_ID:
+        table_id = ip_profile.props.route_table
+    return table_id
+
+
+def _get_running_route_table_id(nm_route):
+    table_id = iplib.KERNEL_MAIN_ROUTE_TABLE_ID
+    table = nm_route.get_attribute('table')
+    if table:
+        table_id = table.get_uint32()
+    return table_id
+
+
+def _get_iface_name(active_connection):
+    """
+    Return interface name for active_connection, return None if error.
+    """
+    # If we have master device, we use it, else we use first device.
```

Comment:
Why is this correct? Could you explain the scenario?
If this is an ac of a bond slave, that slave cannot have any IP on it and therefore routes through it are irrelevant.
Same for a port of a bridge or even a base of a vlan (in the vlan case it is a totally different iface).

Reply:
I will take a look tomorrow.